client/finality-grandpa/src/environment.rs

```diff
@@ -924,22 +929,133 @@ where
+		equivocation: finality_grandpa::Equivocation<Self::Id, Precommit<Block>, Self::Signature>,
 	) {
 		warn!(target: "afg", "Detected precommit equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet
+		match report_equivocation(
+			&self.client.clone(),
+			&self.authority_set.inner().read(),
+			&self.select_chain,
+			equivocation.into(),
+		) {
+			Ok(_) => {}
+			Err(err) => {
+				warn!(target: "afg", "Error reporting precommit equivocation: {:?}", err);
+			}
+		};
 	}
 }
 
+/// Report the given equivocation to the GRANDPA runtime module. This method
+/// generates a session membership proof of the offender and then submits an
+/// extrinsic to report the equivocation. In particular, the session membership
+/// proof must be generated at the block at which the given set was active which
+/// isn't necessarily the best block if there are pending authority set changes.
+fn report_equivocation<BE, Block, Client, SC>(
+	client: &Arc<Client>,
+	authority_set: &AuthoritySet<Block::Hash, NumberFor<Block>>,
+	select_chain: &SC,
+	equivocation: Equivocation<Block::Hash, NumberFor<Block>>,
+) -> Result<(), Error>
+where
+	Block: BlockT,
+	BE: Backend<Block>,
+	Client: crate::ClientForGrandpa<Block, BE>,
+	Client::Api: GrandpaApi<Block>+ SessionMembership<Block, Error = sp_blockchain::Error>,
+	SC: SelectChain<Block> + 'static,
+{
+	let is_descendent_of = is_descendent_of(&**client, None);
+
+	let best_header = select_chain
+		.best_chain()
+		.map_err(|e| Error::Blockchain(e.to_string()))?;
+
+	// block hash and number of the next pending authority set change in the
+	// given best chain.
+	let next_change = authority_set
+		.next_change(&best_header.hash(), &is_descendent_of)
+		.map_err(|e| Error::Safety(e.to_string()))?;
+
+	// find the hash of the latest block in the current set
+	let current_set_latest_hash = match next_change {
+		Some((_, n)) if n.is_zero() => {
+			return Err(Error::Safety(
+				"Authority set change signalled at genesis.".to_string(),
+			))
+		}
+		// the next set starts at `n` so the current one lasts until `n - 1`. if
+		// `n` is later than the best block, then the current set is still live
+		// at best block.
+		Some((_, n)) if n > *best_header.number() => best_header.hash(),
+		Some((h, _)) => {
+			// this is the header at which the new set will start
+			let header = client.header(BlockId::Hash(h))?.expect(
+				"got block hash from registered pending change; \
+				pending changes are only registered on block import; qed.",
+			);
+
+			// its parent block is the last block in the current set
+			*header.parent_hash()
+		}
+		// there is no pending change, the latest block for the current set is
+		// the best block.
+		None => best_header.hash(),
+	};
+
+	// generate membership proof at that block
+	let membership_proof = match client
+		.runtime_api()
+		.generate_session_membership_proof(
+			&BlockId::Hash(current_set_latest_hash),
```

Comment:
Hopefully with the work of @drahnr on #3722  we would be able to call all of this stuff at the most recent block instead, as the offchain DB would hold the set membership tries.